Manifold/Module+Vector.swift

```diff
@@ -0,0 +1,26 @@
+//  Copyright © 2015 Rob Rix. All rights reserved.
+
+extension Module {
+	public static var vector: Module {
+		let Natural: Recur = "Natural"
+		let Vector: Recur = "Vector"
+		let vector = Declaration("Vector",
+			type: .Type --> Natural --> .Type,
+			value: (.Type, Natural, .Type) => { A, n, B in n[.Type, B --> B, Natural => { n in (A --> Vector[A, n] --> B) --> B --> B }] })
```

Comment:
I believe the key here is that the index—the term of type `Natural`—is eliminated in the type, and therefore does not become a part of the value of constructors.

Each constructor can therefore determine its own index: note that `cons`’s index is the successor of its argument’s.

`nil` goes a step further and is always a member of the `zero`-family of `Vector`s. Since `nil`’s eliminator is therefore able to be determined statically, it doesn’t need to mention `Natural` in its value at all.